### ISSUE TITLE
Disabling unreachable code detection partially for the condition completeness check

### ISSUE THREAD

**Issue by falsandtru** [Microsoft/TypeScript#5978]
I want to disable the unreachable code detection for the condition completeness check.

```ts
function f(n) {
  switch(n) {
    case 0: return 0;
    default: return 1;
  }
  // comment base disabling
  // @unreachable
  assert(false); // remove by unassert tools before the release
  throw new Error();
}
```

**Comment by RyanCavanaugh** [comment]
Is there any precedent for this? It seems odd.

If you want a workaround, you can put the `default` section in an `if (true) { ... }`.

**Comment by vladima** [comment]
more like `if (1 == 1) { ... }`, we recognize `true` and `false` in conditions of if-statements and loops when doing reachability checks

**Comment by mhegazy** [comment]
also you can just use `--allowUnreachableCode`

**Comment by falsandtru** [comment]
I cannot find a precedent now. But developers sometimes forget the `default` and `else` clause. This is useful for an annotation. Prevention is better than cure.

other case:

```ts
function f(n) {
  if (n >= 0) {
    return n;
  }
  else if (n < 0) {
    return n;
  }
  else { // don't forget
    return 0;
  }
  // call if you forget the else clause
  throw new Error();
}
```

@mhegazy I want to disable partially.

**Comment by DanielRosenwasser** [comment]
I think this would be better served by a lint tool, but others can weigh in on this.

**Comment by mhegazy** [comment]
I agree with @DanielRosenwasser

**Comment by falsandtru** [comment]
Additionally, a reachability check cannot detect that fault case because an incomplete condition reach the tail statement.

@DanielRosenwasser  But an lint tool cannot allow that code in compilation.

Reachability check disallow that prevention code completely?

**Comment by RyanCavanaugh** [comment]
I find the entire construct here very confusing. You want to make sure that you don't have any unreachable code... so you add a line of unreachable code?

Many coding styles (e.g. ours) say that if you unconditionally `return` in the `true` branch of an `if`, you should *not* have an `else` block (the remainder of the function body is essentially the `else`).

It seems like what you're trying to enforce is "Return on all code paths", which is a reasonable thing for a linter to detect.

**Comment by falsandtru** [comment]
I add oneself a reachability check code instead of reachability check by compiler. This is my use case using a DbC(Design by Contract) like coding style. Is this not good?

https://github.com/falsandtru/arch-stream/blob/v0.0.36/src/lib/message.ts#L126
https://github.com/falsandtru/arch-stream/blob/v0.0.36/src/lib/observable.ts#L33
https://github.com/falsandtru/arch-stream/blob/v0.0.36/src/lib/observable.ts#L79
https://github.com/falsandtru/arch-stream/blob/v0.0.36/src/stream/transform.ts#L68

Assertions will remove by unassert: https://github.com/twada/unassert

**Comment by vladima** [comment]
I guess my question is: in what situation do you expect this assert to be raised? If compiler can statically prove that this code is unreachable (using a very conservative set of checks) then what is the point of putting the bit of code that will never be triggered. The only situation I can imagine when it might be useful is if you want to make your code more resilient to occasional modifications, i.e. if somebody removes a `default` case in switch. In this case compiler can still help you: by using`--noImplicitReturns` you can enforce that all code paths in function should ultimately either return or throw.

**Comment by falsandtru** [comment]
> The only situation I can imagine when it might be useful is if you want to make your code more resilient to occasional modifications, i.e. if somebody removes a default case in switch.

It is right. I feel calm writing that clearly contracted code. I am already using a `--noImplicitReturns`. It is a great evolution but I cannot trust `--noImplicitReturns` completely because TypeScript has not a few bugs. I don't want to lose an infallible means. Embedded codes are a trustworthy method.

**Comment by vladima** [comment]
can you please elaborate: did you step on any issue related to reachability checks or you just suspect that they might exists?

**Comment by falsandtru** [comment]
I did not step on any issue but it need for the condition completeness manual checking. I explain the background:

That my switch statement simulate the overloaded function for pattern matching like a Haskell. Haskell check the typings and the condition completeness. My tail assertion is embedding for the checking of the condition completeness. `--noImplicitReturns` and unreachable code detection covered my needs, but it is unstable. That new features cannot use with manual checking.

In short, must I remove the prevention code immediately to use that new features?

**Comment by aleksey-bykov** [comment]
why won't we ask for exhaustive switch and if/then/else that eliminates a possibility of forgotten cases and elses instead?

**Comment by falsandtru** [comment]
I show another case. please wait.

**Comment by falsandtru** [comment]
`--noImplicitReturns` and unreachable code detection does not work in this case:

expected:

```ts
function f(n) {
	switch (n) {
		case 0:
			return 0;
		default:
			if (n > 0) {
				return n;
			}
			else {
				return -n;
			}
	}
	// assert(false), throw error, or return
	
	function g() {
	}
	// many functions...
	function z() {
	}
}
```

actual:

```ts
function f(n) {
	switch (n) {
		case 0:
			return 0;
		default:
			if (n > 0) {
				return n;
			}
			// else...
	}
	// assert(false), throw error, or return
	
	function g() {
	}
	// many functions...
	function z() {
	}
	
	return n; // bad hidden code
}
```

This is a weak point of `--noImplicitReturns` and unreachable code detection.

**Comment by falsandtru** [comment]
Please review this point, @vladima @RyanCavanaugh @mhegazy @DanielRosenwasser .

I think, unreachable code detection should be an optional feature because **unreachable code detection cannot to be the perfect in JavaScript. JavaScript cannot segregate the main statement block(for detect the tail statement of main flow in function) and the declaration block like a `where` statement of Haskell.** So if possible, unreachable code detection should be able to disable partially.

**Comment by falsandtru** [comment]
Improve to #6009
